### ISSUE TITLE
AOT compilation error in 7.1.0: export 'InjectFlags' was not found in '@angular/core'

### ISSUE THREAD

**Issue by cyrilletuzi** [angular/angular#27251]
# 🐞 bug report

### Affected Package

The issue is caused by package @angular/core

### Is this a regression?

 Yes, the previous version in which this bug was not present was: 7.0.4

### Description

Warning raised during AOT compilation:
```
WARNING in ./src/app/app.component.ngfactory.js 8:0-455
"export 'InjectFlags' (reexported as 'InjectFlags_27') was not found in '@angular/core'
```

The number in `InjectFlags_x` can vary.

## 🔬 Minimal Reproduction

Code reproduction here: [https://stackblitz.com/edit/angular-issue-repro2-hrbbya](https://stackblitz.com/edit/angular-issue-repro2-hrbbya) but the error happens at specific times of the compilation so can't be seen in Stackblitz

Github repro: [https://github.com/cyrilletuzi/ng-aot-issue](https://github.com/cyrilletuzi/ng-aot-issue)

Repros:
* ng new
* update to Angular 7.1 (as CLI still installs 7.0...)
* ng add @angular/material
* import `MatSnackBarModule` in `AppModule`
* inject `MatSnackBar` in `AppComponent` constructor
* ng serve --aot
* Save one or two more times to trigger compilation > error above

Important:
*  it only happens with AOT compilation (so `--aot` is important)
* first compilation is OK, but the next ones raise the warning (so the last step is important)

## 🔥 Exception or Error
<pre><code>
WARNING in ./src/app/app.component.ngfactory.js 8:0-455
"export 'InjectFlags' (reexported as 'InjectFlags_27') was not found in '@angular/core'
</code></pre>

## 🌍  Your Environment

Tried to isolate the issue.

**Angular Version:**

No error with:
* CLI 7.0.6
* Material 7.0.6
* Angular 7.0.4
or with:
* CLI 7.0.6
* Material 7.1.0
* Angular 7.0.4

Error with:
* CLI 7.0.6
* Material 7.0.4
* **Angular 7.1.0**
and:
* CLI 7.0.6
* Material 7.1.0
* **Angular 7.1.0**

So I don't think it's a Material error. `@angular/core` seems to have renamed something used by other packages.

**Comment by alexzuza** [comment]
Looks like the error happens due to dependency to `BidiModule` from `@angular/cdk` which already uses new scope API for injectables https://github.com/angular/material2/blob/df488b0c68847fdef4cd58b9b8dbd0e644cb9b16/src/cdk/bidi/dir-document-token.ts#L28-L36

So you can replace `MatSnackBarModule` with `BidiModule` in `AppModule` and `MatSnackBar` with `Directionality` and catch the same error.

Here's also reproduction without `@angular/material`:

**app.component.ts**

```
import { Component, inject, Inject, InjectionToken } from '@angular/core';
import { DOCUMENT } from '@angular/common';

export const DIR_DOCUMENT = new InjectionToken<Document>('doc', {
  providedIn: 'root',
  factory: DIR_DOCUMENT_FACTORY,
});

export function DIR_DOCUMENT_FACTORY(): Document {
  return inject(DOCUMENT);
}

@Component({
  selector: 'app-root',
  template: 'root template',
})
export class AppComponent {
  constructor(@Inject(DIR_DOCUMENT) doc: Document) {
    console.log(doc);
  }
}
```
https://stackblitz.com/edit/angular-issue-repro2-82qyhc?file=src%2Fapp%2Fapp.component.ts

**Comment by DuncanFaulkner** [comment]
Hi 
Just upgraded to 7.1.0 and now getting these errors, will there be a fix for this?

**Comment by alxhub** [comment]
Thanks @cyrilletuzi for the report!

It appears that an commit between 7.0.4 and 7.1.0 (my suspicion is 2c7386c) changed things around in our import graph enough to trigger `ngc` writing re-exports of injection-related functions into the application `NgFactory` in question. I'll be digging into why this happens exactly, but the logic is non-trivial and sometimes unstable (it can depend on the order in which files are referenced, etc). This is problem #1.

Problem #2 is that as part of writing these re-exports, ngc re-exports `InjectFlags`, a `const enum` with no runtime representation (these are somewhat akin to interfaces, which ngc knows not to re-export). It's very likely that the code in ngc that deals with re-exports predates const enums, and doesn't know to exclude them.

In the meantime, I've submitted an interim fix in #27279, to remove the `const` from `InjectFlags`, which gives it a runtime value. That suppresses the Webpack warning.

The good news is that this problem:

1) has no runtime impact, as nothing reads the re-export, and
2) only occurs in watch mode, on re-compilations with an old `ts.Program`.

Thus, it's completely safe to ignore the warnings.

**Comment by IgorMinar** [comment]
Alex and I reviewed this one more time and agreed that using `--aot` is not a uncommon use-cases and seeing this warning in the development mode is highly annoying. We have a workaround that will be implemented for the next patch release.
